File: algorithms/groupAnagrams/group-anagrams.py
# https://leetcode.com/problems/group-anagrams/description/

# Given an array of strings, group anagrams together.

import collections
class Solution(object):
    def groupAnagrams(self, strs):
        """
            :type strs: List[str]
             :rtype: List[List[str]]
             """
         
        anagrams_map, result = collections.defaultdict(list), []        
        for s in strs:
           sorted_str = ("").join(sorted(s))
           anagrams_map[sorted_str].append(s)
        for anagram in anagrams_map.values():
            anagram.sort()
            result.append(anagram)
        return result


if __name__ == "__main__":
    result = Solution().groupAnagrams(["cat", "dog", "act", "mac"])
    print result

File: algorithms/groupAnagrams/group-anagrams.cpp
//URL:https://leetcode.com/problems/group-anagrams/description/
/*Given an array of strings, group anagrams together.
 *
 * For example, given: ["eat", "tea", "tan", "ate", "nat", "bat"], 
 * Return:
 *
 * [
 *   ["ate", "eat","tea"],
 *   ["nat","tan"],
 *   ["bat"]
 * ]
 *
 * Note: All inputs will be in lower-case.
 *
 **/


//#include<stdio.h>
#include<iostream>
#include<vector>
#include<string>
#include<algorithm>
#include<unordered_map>

using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string> & strs){
        unordered_map<string, vector<string>> groups;
        for (const auto& str : strs) {
            string tmp{str};
            sort(tmp.begin(),tmp.end());
            groups[tmp].emplace_back(str);
        }

        vector<vector<string>> anagrams;
        for (const auto& kvp : groups) {
            vector<string> group;
            for (const auto& str : kvp.second) {
                group.emplace_back(str);
            } 
            sort(group.begin(), group.end());
            anagrams.emplace_back(move(group));
        }

        return anagrams;
    }
};

/*
int main() {
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    Solutions s;
    auto ans = s.groupAnagrams(strs);

    for(int i = 0; i < ans.size(); i++) {
        for (int j = 0; j< ans[i].size(), j++) {
            cout << ans[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}
*/
